// gitart-scroll-carousel/dist/GSLayoutNumeric-93bbaef6.mjs
import { defineAsyncComponent as m, defineComponent as p, computed as _, resolveComponent as f, openBlock as t, createElementBlock as n, createElementVNode as s, renderSlot as i, createBlock as l, createCommentVNode as o, createTextVNode as u, toDisplayString as a, pushScopeId as y, popScopeId as v } from "vue";
import { _ as b } from "./GSArrow-d3dc6242.mjs";
const g = m(() => import("./GSArrow-d3dc6242.mjs").then((e) => e.G)), S = p({
  name: "GSLayoutNumeric",
  components: {
    GSArrow: g
  },
  props: {
    onMove: {
      type: Function,
      required: !0
    },
    disabledSide: {
      type: String,
      required: !0
    },
    currentItem: {
      type: Number,
      required: !0
    },
    currentItemByPercent: {
      type: Number,
      required: !0
    },
    items: {
      type: Array,
      required: !0
    },
    /**
     * layout specific props
     */
    disableArrows: {
      type: Boolean,
      default: !1
    },
    disableCounter: {
      type: Boolean,
      default: !1
    },
    percentCounter: {
      type: Boolean,
      default: !1
    }
  },
  setup(e) {
    return {
      counter: _(() => (e.percentCounter ? e.currentItemByPercent : e.currentItem) + 1)
    };
  }
});
const w = (e) => (y("data-v-2fd0d2f6"), e = e(), v(), e), C = { class: "gsl-numeric" }, k = { class: "gsl-numeric__indicator" }, h = {
  key: 0,
  class: "gsl-numeric__contorls"
}, A = {
  key: 1,
  class: "gsl-numeric__contorls-text"
}, B = /* @__PURE__ */ w(() => /* @__PURE__ */ s("span", null, " / ", -1));
function G(e, r, I, N, $, q) {
  const d = f("GSArrow");
  return t(), n("div", null, [
    s("div", C, [
      i(e.$slots, "track", {}, void 0, !0)
    ]),
    s("div", k, [
      i(e.$slots, "indicator", {}, void 0, !0)
    ]),
    e.disableArrows && e.disableCounter ? o("", !0) : (t(), n("div", h, [
      e.disableArrows ? o("", !0) : (t(), l(d, {
        key: 0,
        disabled: e.disabledSide === "left",
        side: "left",
        class: "gsl-numeric__arrow gsl-numeric__arrow--left",
        onClick: r[0] || (r[0] = (c) => e.onMove(-1))
      }, null, 8, ["disabled"])),
      e.disableCounter ? o("", !0) : (t(), n("div", A, [
        u(a(e.counter) + " ", 1),
        B,
        u(" " + a(e.items.length), 1)
      ])),
      e.disableArrows ? o("", !0) : (t(), l(d, {
        key: 2,
        disabled: e.disabledSide === "right",
        side: "right",
        class: "gsl-numeric__arrow gsl-numeric__arrow--right",
        onClick: r[1] || (r[1] = (c) => e.onMove(1))
      }, null, 8, ["disabled"]))
    ]))
  ]);
}
const V = /* @__PURE__ */ b(S, [["render", G], ["__scopeId", "data-v-2fd0d2f6"]]);
export {
  V as G
};
